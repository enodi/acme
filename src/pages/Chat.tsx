import React, { useState, useEffect } from "react";

import Sidebar from "../components/Sidebar";
import ChatScreen from "../components/ChatScreen";
import Header from "../components/Header";
import data from "../MOCK_DATA.json";
import { userDataFormat } from "../models";

const Chat: React.FC = () => {
  const [favouriteConversation, setFavouriteConversation] = useState(
    "conversations"
  );
  const [userData, setUserData] = useState<any>([]);
  const [messageText, setMessageText] = useState("");
  const [toggle, setToggle] = useState(false);
  const [selectedItem, setSelectedItem] = useState<any>({});
  const [messages, setMessages] = useState<any>([]);
  const [conversation, setConversation] = useState<any>({});

  useEffect(() => {
    setUserData(data);
  }, []);

  const filterFavourite = (event: React.ChangeEvent<HTMLSelectElement>) => {
    if (event.target.value === "isFavorite") {
      setFavouriteConversation(event.target.value);
      const favourites = userData.filter(
        (item: userDataFormat) => item.isFavorite === true
      );
      setUserData(favourites);
    } else {
      setFavouriteConversation(event.target.value);
      setUserData(data);
    }
  };

  const filterByFullname = (event: React.ChangeEvent<HTMLInputElement>) => {
    const filteredData = data.filter((item) => {
      const firstName = item.firstName.toLowerCase();
      const lastName = item.lastName.toLowerCase();

      return (
        `${firstName} ${lastName}`.indexOf(event.target.value.toLowerCase()) !==
        -1
      );
    });

    setUserData(filteredData);
  };

  const toggleConversationType = (
    event: React.MouseEvent<SVGSVGElement>,
    id: number
  ) => {
    event.stopPropagation();
    const updatedData: userDataFormat[] = [];
    userData.reduce((acc: {}, item: userDataFormat) => {
      if (item.id === id) {
        item.isFavorite = !item.isFavorite;
      }
      updatedData.push(item);
    }, {});
    setUserData(updatedData);
  };

  const onTextChange = (event: React.ChangeEvent<HTMLInputElement>) =>
    setMessageText(event.target.value);

  const toggleSidebar = () => setToggle(!toggle);

  const hideOverlay = () => setToggle(false);

  const handleSelected = (selected: userDataFormat) => {
    const isSelected = false;
    const updatedData: userDataFormat[] = [];
    setMessages([]);
    setConversation({});

    userData.reduce((acc: {}, item: userDataFormat) => {
      if (selected.id === item.id) {
        item.isSelected = !isSelected;
        setSelectedItem(item);
      } else {
        item.isSelected = isSelected;
      }
      updatedData.push(item);
    }, {});
    setUserData(updatedData);
  };

  const handleKeyDown = (event: any) => {
    const chat: string[] = [];
    if (event.key === "Enter") {
      chat.push(event.target.value);
      setMessages([...messages, ...chat]);
      selectedItem.messages = [...messages, ...chat];
      setConversation(selectedItem);
      setMessageText("");
    }
  };

  return (
    <div className="wrapper">
      <Header toggleSidebar={toggleSidebar} selectedItem={selectedItem} />
      <div className="chat">
        {toggle && <div className="overlay" onClick={hideOverlay} />}
        <Sidebar
          userData={userData}
          toggleConversationType={toggleConversationType}
          filterByFullname={filterByFullname}
          filterFavourite={filterFavourite}
          favouriteConversation={favouriteConversation}
          toggle={toggle}
          handleSelected={handleSelected}
        />
        <ChatScreen
          selectedItem={selectedItem}
          conversation={conversation}
          handleKeyDown={handleKeyDown}
          onTextChange={onTextChange}
          messageText={messageText}
        />
      </div>
    </div>
  );
};

export default Chat;
