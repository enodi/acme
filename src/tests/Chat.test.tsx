import React from "react";
import { shallow } from "enzyme";

import Chat from "../pages/Chat";

let wrapper: any;
beforeEach(() => {
  wrapper = shallow(<Chat />);
});

test("should render Chat component correctly", () => {
  expect(wrapper).toMatchSnapshot();
});
