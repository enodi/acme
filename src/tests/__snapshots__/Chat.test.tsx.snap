// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Chat component correctly 1`] = `
<div
  className="wrapper"
>
  <Header
    selectedItem={Object {}}
    toggleSidebar={[Function]}
  />
  <div
    className="chat"
  >
    <Sidebar
      favouriteConversation="conversations"
      filterByFullname={[Function]}
      filterFavourite={[Function]}
      handleSelected={[Function]}
      toggle={false}
      toggleConversationType={[Function]}
      userData={Array []}
    />
    <ChatScreen
      conversation={Object {}}
      handleKeyDown={[Function]}
      messageText=""
      onTextChange={[Function]}
      selectedItem={Object {}}
    />
  </div>
</div>
`;
